### ISSUE TITLE
F1 score > 1

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [PyTorchLightning/pytorch-lightning#5662]
## 🐛 Bug

F1 is > 1

From comet:

![image](https://user-images.githubusercontent.com/15908060/105865655-6d471a00-5ff3-11eb-801e-cfface29fc81.png)

Thank you!

Cheers,

Francesco

## Please reproduce using the BoringModel

```python
class MySystem(pl.LightningModule):

    def __init__(self, model: nn.Module, criterion: Callable = nn.CrossEntropyLoss, lr :float = 1e-3):
        super().__init__()
        self.model = model
        self.criterion = criterion()
        self.lr = lr
        self.metrics = nn.ModuleDict({
            'accuracy': Accuracy(),
            'f1': F1()
        })
        # this apis are so bad
        
    def compute_metrics(self, out: Tensor, batch: Union[Tensor, Tensor]) -> Dict[str, Tensor]:
        x, y = batch
        metrics = {}
        for name, metric in self.metrics.items():
            metrics[name] = metric(out, y)
        return metrics

    def basic_step(self, batch: Union[Tensor, Tensor], name_space: str) -> Tensor:
        x, y = batch
        out = self.model(x)
        loss = self.criterion(out, y)
        metrics = self.compute_metrics(out, batch)
        self.log_dict({f'{name_space}_loss': loss, **
                       {f'{name_space}_{k}': v for k, v in metrics.items()}})
        return loss

    def training_step(self, batch: Union[Tensor, Tensor], batch_idx: int) -> Tensor:
        loss = self.basic_step(batch, name_space='train')
        return loss

    def validation_step(self, batch: Union[Tensor, Tensor], batch_idx: int) -> Tensor:
        loss = self.basic_step(batch, name_space='val')
        return loss

    def test_step(self, batch: Union[Tensor, Tensor], batch_idx: int) -> Tensor:
        loss = self.basic_step(batch, name_space='test')
        return loss

    def configure_optimizers(self):
        optimizer = optim.AdamW(self.parameters(), lr=self.lr)
        scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer)
        return {
            'optimizer': optimizer,
            'lr_scheduler': scheduler,
            'monitor': 'val_loss'
        }

```

The training is correct as loss and accuracy are consistent, 

```python
...
    # our callbacks
    callbacks = [
        ModelCheckpoint(
            monitor='val_loss',
            dirpath=model_checkpoint_dir,
            filename='{epoch:02d}-{val_loss:.2f}'
        ),
        EarlyStopping(monitor='val_loss', patience=10, verbose=True)
    ]
    # using commet
    logger = CometLogger(
        api_key=secrets['COMET_API_KEY'],
        project_name="test",
        workspace="francescosaveriozuppichini"
    )
    logger.log_hyperparams(params)
    
    system = MySystem(model=cnn, lr=params['lr'])

    trainer = Trainer(gpus=1, min_epochs=params['epochs'],
                      progress_bar_refresh_rate=20, logger=logger,
                      callbacks=callbacks)

    trainer.fit(system, train_dl, val_dl)

    print(trainer.test(test_dataloaders=test_dl))

```

### Environment

* CUDA:
        - GPU:
                - GeForce GTX 1080 Ti
        - available:         True
        - version:           11.0
* Packages:
        - numpy:             1.18.4
        - pyTorch_debug:     True
        - pyTorch_version:   1.7.0+cu110
        - pytorch-lightning: 1.1.5
        - tqdm:              4.46.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.2
        - version:           #45~20.04.1-Ubuntu SMP Fri Jan 15 11:35:04 UTC 2021

[Edit] Forgot to say thank you :)

**Comment by FrancescoSaverioZuppichini** [comment]
btw I know you have a very low code coverage of only 89% but I was wondering how the metrics are tested, I cannot use and suggest pl to friends if I am not 100% sure the metrics are 100% tested (and it should be easy to do IHMO)

**Comment by teddykoker** [comment]
Hi Francesco,

The full lightning code base registers at < 90% coverage only because of some accelerator related methods that we must test externally on other hardware. The full metrics package is rigourously tested against sklearn, see: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/metrics/classification/test_f_beta.py for the FBeta tests.

If you could provide an example input that provides an incorrect result that would be very helpful in diagnosing this issue.

**Comment by SkafteNicki** [comment]
@FrancescoSaverioZuppichini could you try initializing the `F1` metric with the `num_classes` argument set correctly.
I can reproduce error by not setting the argument (as in your code)
``` python
metric = F1()
val = metric(torch.randint(0, 10, (1000,)), torch.randint(0, 10, (1000,)))
print(val) # val can be over 1

metric = F1(num_classes=10)
val = metric(torch.randint(0, 10, (1000,)), torch.randint(0, 10, (1000,)))
print(val)  # val should be below 1
```

**Comment by FrancescoSaverioZuppichini** [comment]
> Hi Francesco,
> 
> The full lightning code base registers at < 90% coverage only because of some accelerator related methods that we must test externally on other hardware. The full metrics package is rigourously tested against sklearn, see: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/metrics/classification/test_f_beta.py for the FBeta tests.
> 
> If you could provide an example input that provides an incorrect result that would be very helpful in diagnosing this issue.

Thank you for pointing out the test file. Very good to hear that all the metrics are tested :) Setting the `num_classes` to the class number fixed the issue! Thank you @SkafteNicki for the blazing fast reply. 

I think we can close it, who wants to do the honours?

Cheers,

Francesco

**Comment by SkafteNicki** [comment]
@FrancescoSaverioZuppichini I have open PR #5663 which removes the default value of 1 from `F1` metric so this kind of error hopefully should not happen in the future (as the user will be forced to provide the arg). When the PR is merged, this issue should auto close :]

**Comment by FrancescoSaverioZuppichini** [comment]
Cool, thank you again @SkafteNicki . Take care and keep up the good work! :)

**Comment by noamzilo** [comment]
> 
> 
> @FrancescoSaverioZuppichini I have open PR #5663 which removes the default value of 1 from `F1` metric so this kind of error hopefully should not happen in the future (as the user will be forced to provide the arg). When the PR is merged, this issue should auto close :]

I was not forced to provide `num_classes` on 1.1.5, which brought me here.

was it decided it was better not to force this?
